### PR TITLE
Dask dataframes for parallelism and handling larger datasets

### PR DESCRIPTION
**Pull request by George3d6**
# DO NOT MERGE

Only opening this in order to have an easier way to look at difs,

### CODE REVIEW

File: mindsdb/libs/controllers/transaction.py [hunk 1]
```diff
@@ -171,7 +171,7 @@ def _execute_learn(self):
             self._call_phase_module(clean_exit=True, module_name='DataSplitter')
 
             self._call_phase_module(clean_exit=True, module_name='DataTransformer', input_data=self.input_data)
-
+            exit()
```
Comment (torrmal) [inline mindsdb/libs/controllers/transaction.py]: ?

Comment (torrmal) [reply mindsdb/libs/controllers/transaction.py]: @George3d6 ^

File: mindsdb/libs/phases/data_extractor/data_extractor.py [hunk 2]
```diff
@@ -70,20 +70,25 @@ def _get_prepared_input_df(self):
...
+            # @TODO This gets converted to a pandas Datafram from a mindsdb DataSource , not sure why
+            df = df.where((pd.notnull(df)), None)
 
         # if this is a predict statement, create use model_when_conditions to shape the dataframe
         if  self.transaction.lmd['type'] == TRANSACTION_PREDICT:
             if self.transaction.hmd['when_data'] is not None:
                 df = self.transaction.hmd['when_data']
-                df = df.where((pandas.notnull(df)), None)
+                # @TODO This gets converted to a pandas Datafram from a mindsdb DataSource here, not sure why
+                df = df.where((pd.notnull(df)), None)
```
Comment (torrmal) [inline mindsdb/libs/phases/data_extractor/data_extractor.py]: we need to take out the null rows

File: mindsdb/libs/phases/data_extractor/data_extractor.py [hunk 3]
```diff
@@ -70,20 +70,25 @@ def _get_prepared_input_df(self):
...
             elif self.transaction.hmd['model_when_conditions'] is not None:
 
                 # if no data frame yet, make one
                 df = self._get_data_frame_from_when_conditions()
 
-
+        df = df.reset_index()
+        df = dd.from_pandas(df, npartitions=1)
+        #df = df.reset_index().compute()
+        print(type(df))
```
Comment (torrmal) [inline mindsdb/libs/phases/data_extractor/data_extractor.py]: ?

File: mindsdb/libs/phases/data_extractor/data_extractor.py [hunk 4]
```diff
@@ -113,7 +119,7 @@ def _validate_input_data_integrity(self):
         """
         :return:
         """
-        if self.transaction.input_data.data_frame.shape[0] <= 0:
+        if len(self.transaction.input_data.data_frame) <= 0:
```
Comment (torrmal) [inline mindsdb/libs/phases/data_extractor/data_extractor.py]: why this change?

Comment (George3d6) [reply mindsdb/libs/phases/data_extractor/data_extractor.py]: Because initially dask was being used in this part (switched back to pandas in the last commit) and it doesn't support `.shape`

File: mindsdb/libs/phases/data_extractor/data_extractor.py (hunk 3 above)
Comment (George3d6) [reply mindsdb/libs/phases/data_extractor/data_extractor.py]: This is just debugging code

File: mindsdb/libs/phases/data_extractor/data_extractor.py (hunk 2 above)
Comment (George3d6) [reply mindsdb/libs/phases/data_extractor/data_extractor.py]: no change was made here other than renaming `pd` to `pandas`

File: mindsdb/libs/controllers/transaction.py (hunk 1 above)
Comment (George3d6) [reply mindsdb/libs/controllers/transaction.py]: Debugging code, not supposed to me merged.

**Comment by George3d6** [comment]
Closing this, maybe consider migrating to : https://vaex.readthedocs.io/en/latest/ instead of dask, seems to be better optimized than pandas but meant for datasets in the dozens or low hundreds of GBs, which is closer to what we want (dask seems meant for bigger datasets).
